### ISSUE TITLE
stop simulations that have died out

### ISSUE THREAD

**Issue by rjwalters** [rulehunt/rulehunt#70]
don't bother continuing to simulate if there are no active tiles

**Comment by rjwalters** [comment]
## Implementation Guidance

This feature automatically pauses simulations that have died out (no living cells or no activity) to conserve CPU resources and signal to the user that the simulation is complete.

## Acceptance Criteria

- [ ] Detect "died out" state (population ≈ 0 or activity = 0 for N consecutive steps)
- [ ] Auto-pause simulation when died-out state is detected
- [ ] Visual indication that simulation died out (not just paused by user)
- [ ] Highlight/pulse the ♻️ reset button to draw attention (per @irgolic's suggestion)
- [ ] Optional: Show toast/notification: "Simulation died out"
- [ ] Works on both mobile and desktop layouts
- [ ] User can still manually resume if desired (button changes to "Play" state)
- [ ] Dark mode support for visual indicators

## Technical Approach

**Detection Logic:**

Use existing `StatisticsTracker` data to detect die-out. A simulation has "died out" when:
1. **Population-based**: `population / gridArea < 0.001` (< 0.1% alive)
2. **Activity-based**: `activity === 0` for N consecutive steps (no changes)

**Option A: Population threshold (Simpler, recommended)**
```typescript
// In statistics.ts or CA engine
function isDiedOut(stats: GridStatistics, gridArea: number): boolean {
  const populationRatio = stats.population / gridArea
  return populationRatio < 0.001 // Less than 0.1% alive
}
```

**Option B: Activity-based (More robust)**
```typescript
// Track consecutive zero-activity steps
let zeroActivityCount = 0
const DIED_OUT_THRESHOLD = 10 // 10 steps with no activity

function checkDiedOut(stats: GridStatistics): boolean {
  if (stats.activity === 0) {
    zeroActivityCount++
  } else {
    zeroActivityCount = 0
  }
  return zeroActivityCount >= DIED_OUT_THRESHOLD
}
```

**Recommendation**: Use **Option A** for initial implementation (simpler, catches complete die-out). Add Option B in v2 if users report false positives.

**Auto-pause Integration:**

In the CA play loop (e.g., `cellular-automata-base.ts:384`), add check after each step:
```typescript
playInterval = window.setInterval(() => {
  this.step(active_ruleset)
  
  // Check for die-out
  const stats = this.statisticsTracker?.getRecentStats(1)[0]
  if (stats && this.isDiedOut(stats)) {
    this.pause()
    this.onDiedOut?.() // Callback to notify UI
  }
}, intervalMs)
```

**UI Indication:**

Mobile layout (pulse reset button):
```typescript
// In mobile.ts or similar
function onSimulationDiedOut() {
  resetButton.classList.add('pulse-animation')
  resetButton.style.borderColor = 'orange'
  // Optional: show toast
  showToast('Simulation died out - tap ♻️ to restart')
}

// CSS animation (add to styles)
@keyframes pulse {
  0%, 100% { transform: scale(1); }
  50% { transform: scale(1.1); }
}
.pulse-animation {
  animation: pulse 1s ease-in-out infinite;
}
```

Desktop layout:
```typescript
// Similar approach in desktop/simulation.ts
function onSimulationDiedOut() {
  btnReset.classList.add('animate-pulse') // Use Tailwind's built-in pulse
  btnReset.style.borderColor = '#f97316' // Orange border
}
```

## Files to Modify

**Core CA Engine:**
- `src/cellular-automata-base.ts:374-396` - Modify `play()` method to check for die-out
  - Add `onDiedOut` callback property
  - Check statistics after each step
  - Auto-pause when died out

**Statistics Tracker:**
- `src/statistics.ts:1-507` - Add `isDiedOut()` helper method
  - Accept threshold as parameter (default 0.001)
  - Return boolean

**Mobile Layout:**
- `src/components/mobile/layout.ts` or similar - Add die-out visual feedback
  - Register `onDiedOut` callback
  - Pulse reset button
  - Optional: show toast notification

**Desktop Layout:**
- `src/components/desktop/simulation.ts:1-100` - Add die-out visual feedback
  - Register `onDiedOut` callback
  - Highlight reset button

## Test Plan

**Manual Testing:**

1. **Complete Die-out (empty grid):**
   - Load a rule that dies out completely (e.g., all cells die)
   - Play simulation → verify it auto-pauses when population reaches 0
   - Verify reset button pulses/highlights
   - Click reset → verify animation stops, simulation restarts

2. **Static Pattern (no activity):**
   - Load a rule that reaches stable state (e.g., still life)
   - Play simulation → verify it auto-pauses after N steps of no activity
   - Verify visual feedback

3. **Complex Pattern (should NOT trigger):**
   - Load a complex rule with ongoing activity
   - Play simulation for 100+ steps
   - Verify it does NOT auto-pause

4. **Manual Pause:**
   - Pause simulation before it dies out
   - Verify reset button does NOT highlight (only on auto-pause)

5. **Resume After Die-out:**
   - Let simulation die out (auto-pause)
   - Click Play button → verify it resumes (even though grid is dead)
   - User can still manually control

6. **Mobile & Desktop:**
   - Test on both layouts
   - Verify consistent behavior

7. **Dark Mode:**
   - Toggle dark mode
   - Verify pulse animation and border color visible in both themes

**Edge Cases:**
- Very small population but still active (< 0.1% but changing) → should NOT pause
- Oscillating pattern (activity > 0) → should NOT pause
- User pauses then resumes → should NOT highlight reset button
- Rapid die-out (< 10 steps) → should still detect

## Dependencies

None - uses existing statistics tracking infrastructure.

## Performance Considerations

- Die-out check is O(1) (just check population/activity values)
- No additional computation beyond existing statistics tracking
- Auto-pause reduces CPU usage significantly for died-out sims

## Future Enhancements

- User preference to disable auto-pause (for studying dead states)
- Different thresholds for "died out" vs "stabilized" (still life detection)
- Analytics: track % of sims that die out vs stabilize vs run complex
- Auto-restart option: automatically load next rule when one dies out (data generation mode)

## Similar Features

- Conway's Game of Life simulators often have "stable pattern detection"
- Golly (CA tool) has pattern analysis and stability detection
- Mobile games often highlight "next action" buttons with pulse animations
